src/backend/empty/src/memory.rs

```diff
@@ -0,0 +1,140 @@
+use hal::{
+    device::{AllocationError, MapError, OutOfMemory},
+    memory::Segment,
+    MemoryTypeId,
+};
+use std::cell::UnsafeCell;
+use std::convert::TryInto;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A memory allocation.
+#[derive(Debug)]
+pub struct Memory {
+    /// The type of memory this block was allocated from.
+    memory_type: MemoryTypeId,
+    /// Size of the allocated memory block.
+    size: u64,
+    /// The backing memory allocation.
+    data: UnsafeCell<Box<[u8]>>,
+    /// Marker for determining if data is currently mapped.
+    mapped: AtomicBool,
+}
+
+impl Memory {
+    pub fn allocate(memory_type: MemoryTypeId, size: u64) -> Result<Self, AllocationError> {
+        assert_eq!(memory_type.0, 0, "We only support one memory type");
+
+        // TODO: track how many allocations are made and set a limit
+
+        // Allocate a buffer in RAM
+        let data = {
+            let size = size
+                .try_into()
+                // If we're on 32-bit and the given size is greater than 2^32,
+                // we certainly can't allocate it.
+                .map_err(|_| AllocationError::OutOfMemory(OutOfMemory::Host))?;
+
+            vec![0u8; size].into_boxed_slice()
+        };
+
+        let memory = Memory {
+            memory_type,
+            size,
+            data: UnsafeCell::new(data),
+            mapped: AtomicBool::new(false),
+        };
+        Ok(memory)
+    }
+
+    pub fn map(&self, segment: Segment) -> Result<*mut u8, MapError> {
+        if segment.offset >= self.size {
+            return Err(MapError::OutOfBounds);
+        }
+        if let Some(size) = segment.size {
+            if segment.offset + size > self.size {
+                return Err(MapError::OutOfBounds);
+            }
+        }
+
+        let order = Ordering::SeqCst;
+        self.mapped
+            .compare_exchange(false, true, order, order)
+            .expect("Cannot map memory twice");
+
+        let data = unsafe { &mut *self.data.get() };
+        Ok(data.as_mut_ptr())
+    }
+
+    pub fn unmap(&self) {
+        let order = Ordering::SeqCst;
+        self.mapped
+            .compare_exchange(true, false, order, order)
+            .expect("Cannot unmap memory that isn't mapped");
+    }
+}
+
+impl Drop for Memory {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            let order = Ordering::SeqCst;
+            self.mapped
+                .compare_exchange(false, false, order, order)
+                .expect("Cannot free memory if still mapped");
+        }
+    }
+}
+
+unsafe impl Sync for Memory {}
+
+#[cfg(test)]
+mod tests {
```

Comment:
we don't actually need to test this. Gfx-hal backends are inherently unsafe, they are free to assume the valid usage rules of Vulkan are followed.

Reply:
Yep, I understand, removed them now